CIPs/cip-0044.md

```diff
@@ -0,0 +1,76 @@
+
+This CIP specifies the parameters and algorithms for left/right constrained PRFs for use in Celo protocols.
+
+## Motivation
+This CIP introduces left/right constrained PRFs that can be used to develop improved privacy-protection mechanisms for users on the Celo blockchain.
+
+## Specification
+We follow the specification of (left/right) constrained PRFs as introduced in [BW13](BW13).
+
+### Constrained PRFs
+Given a key space `K`, a domain `X`, range `Y`, let `F` be a pseudorandom function that takes as inputs `(k, x)` for `k` in `K` and `x` in `X` and returns `F(k, x)` in `Y`. 
+
+We say that `F` is constrained with respect to a set of constraints `S` if there is a new key space `K_c`, such that we can define the following functions:
+- `F.constrain(k, S)` which provides a constrained key `k_S` in `K_c`. This key `k_S` enables the evaluation of `F(k, x)` for all `x` in `S` and no other `x`.
+- `F.eval(k_S, x)` which outputs `F(k, x)` if `x` is in `S` and `nil` otherwise.
+
+In practice, `S` is usually a set of predicates `PP` such that for any `p` in `PP`, we have `F.eval(k_p, x) = F(k, x)` if `p(x) = 1` and `F.eval(k_p, x) = nil` otherwise.
+
+### Left/right constrained PRFs
+Let `F: K x X^2 -> Y` be a PRF. Then, `F` is a left/right constrained PRF if `F` is constrained by the following predicates `P_LR = {(p_wL, p_wR), w in X}` where `p_wL(x,y) = 1` iff `x = w` and `p_wR(x,y) = 1` iff `y = w`. Left/right constrained PRFs have the property such that for all points in `(w, y)` in `X^2`, `F(k, (w, y))` can be evaluated at all the points in which the left side is fixed to `w`. Similarly for right constrained PRFs. 
+
+We now instantiate a left/right constrained PRF based on bilinear pairings to be used in the Celo ecosystem.
+
+Let `K` be `Z_p`, the set of integers modulo `p`, and let `(G_1, G_2, G_T)` be the pairing-friendly curve BLS12-377 with optimal ate bilinear-pairing operation `e: G_1 x G_2 -> G_T` as defined in [CIP30](CIP30). 
+
+Then we define `F(k, (x,y)) := e(H_1(x), H_2(y))^k`, where
+- `H_1()` is the hash to `G_1` function defined in [HASH_TO_CURVE](HASH_TO_CURVE),
+- `H_2()` is the hash to `G_2` function defined in [HASH_TO_CURVE](HASH_TO_CURVE),
+- `k` is a secret key chosen uniformly at random from `K`.
+
+The constrained keys for the predicates `p_wL` and `p_wR` are defined as: `k_x* = H_1(x*)^k` and `k_y* = H_2(y*)^k` where `(x*, y*)` is in `X^2`.
```

Comment:
Writing out `F.constrain` and `F.eval` here would be good instead of leaving the reader to fill in the blanks.

Reply:
I don't think this is necessary simply because we have clearly described how to construct `F` and how to construct constrained keys.